### PR TITLE
41 json implement discriminator logic

### LINKED ISSUE #41: [JSON] Implement Discriminator logic

### PR DESCRIPTION
**Pull request by jp06upel**
closes #41

### CODE REVIEW

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 1]
```diff
@@ -15,115 +15,186 @@
...
+            // Create new Parser instance with new PropertyResolver for scope of this schema only
+            JsonSchemaParser jsonSchemaParser = new JsonSchemaParser(resolver, store, new PropertyResolver(store));
+            return jsonSchemaParser.parseSchema(schemaEntry.getValue(), schemaEntry.getKey()); // Parse with new Parser instance
+        }).collect(Collectors.toList());
     }
 
-    public SchemaDescriptor parseSchema(Schema<?> schema, String name){
+    private SchemaDescriptor parseSchema(Schema<?> schema, String name){
 
         SchemaDescriptor schemaDescriptor = resolver.resolve(SCHEMA_REFSTRING + name);
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:59]: if schema already exists, is the rest of this function necessary (just overwrites the existing schema with same schema isnt it) ?

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 2]
```diff
@@ -15,115 +15,186 @@
...
+            }
+        } else if (schema.getOneOf() != null && !schema.getOneOf().isEmpty()){
+            for (Schema x : schema.getAllOf()){
+                schemaDescriptor.getOneOfSchemas().add(parseXof(x));
+            }
+        } else if (schema.getAnyOf() != null && !schema.getAnyOf().isEmpty()){
+            for (Schema x : schema.getAllOf()){
+                schemaDescriptor.getAnyOfSchemas().add(parseXof(x));
+            }
+        } else {
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:75]: empty checks are already done by the for loops i think. so in theory we can lose these and decrease the count of edge cases which need to be covered by tests

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 3]
```diff
@@ -15,115 +15,186 @@
...
+            schemaDescriptor.setDiscriminator(parseDiscriminator(schema));
+        }
 
         return schemaDescriptor;
 
     }
 
-    private PropertyDescriptor parseProperty(Schema<?> property, String name){
-        PropertyDescriptor propertyDescriptor;
+    private DiscriminatorDescriptor parseDiscriminator(Schema<?> schema) {
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:87]: passing of whole schema could be reduced to passing only discriminator object

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 4]
```diff
@@ -15,115 +15,186 @@
...
+        DiscriminatorDescriptor discriminatorDescriptor = store.create(DiscriminatorDescriptor.class);
+
+        if (schema.getDiscriminator().getPropertyName() != null)
+            discriminatorDescriptor.setProperty(propertyResolver.resolve(schema.getDiscriminator().getPropertyName()));
+
+        if (schema.getDiscriminator().getMapping() != null && !schema.getDiscriminator().getMapping().isEmpty()) {
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:93]: explicit empty check needed?

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 5]
```diff
@@ -15,115 +15,186 @@
...
         if(schema.getExternalDocs() != null)
             schemaDescriptor.setExternalDocs(ExternalDocsParser.parseOne(schema.getExternalDocs(), store));
 
-        schemaDescriptor.setObject(parseProperty(schema, name));
+        if (schema.getAllOf() != null && !schema.getAllOf().isEmpty()){
+            for (Schema x : schema.getAllOf()){
+                schemaDescriptor.getAllOfSchemas().add(parseXof(x));
+            }
+        } else if (schema.getOneOf() != null && !schema.getOneOf().isEmpty()){
+            for (Schema x : schema.getAllOf()){
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:68]: loop over `schema.getAllOf` intended or should we rather loop over `schema.getOneOf`?

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 6]
```diff
@@ -15,115 +15,186 @@
...
+                schemaDescriptor.getOneOfSchemas().add(parseXof(x));
+            }
+        } else if (schema.getAnyOf() != null && !schema.getAnyOf().isEmpty()){
+            for (Schema x : schema.getAllOf()){
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:72]: loop over `schema.getAllOf` intended or should we rather loop over `schema.getAnyOf`?

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 7]
```diff
@@ -15,115 +15,186 @@
...
         if (property == null)
-            return null; // No property present
+            return null; // No Type present
 
         if (property.getType() == null) {
             if (property.get$ref() != null)
                 return parseReference(property.get$ref());
             else
-                return null; // No property present
+                return parseNull(property); // No Type present
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:116]: could pull return one indentation to the left and lose the else (no idea if that is better or worse)

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 8]
```diff
@@ -15,115 +15,186 @@
...
+                typeDescriptor = parseObject(property);
                 break;
-            case StringPropertyDescriptor.TYPE_NAME:
-                propertyDescriptor = parseString(property);
+            case StringTypeDescriptor.TYPE_NAME:
+                typeDescriptor = parseString(property);
                 break;
             default:
-                throw new UnknownTypeException("Unknown Type \"" + property.getType() + "\" in Schema: " + name);
+                throw new UnknownTypeException("Unknown Type \"" + property.getType() + "\"");
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:142]: Didn't we want to create a descriptor of unknown type and throw warning?

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 9]
```diff
@@ -135,53 +206,57 @@ List<EnumValueDescriptor> parseEnumValues(List<?> enums){
...
-        if (schema.getDescription() != null)
-            propertyDescriptor.setDescription(schema.getDescription());
-
-        return propertyDescriptor;
+    private TypeDescriptor parseXof(Schema x){
+        if (x.get$ref() != null) {
+            return parseReference(x.get$ref());
+        } else {
+            return parseType(x);
+        }
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:251]: could lose the brackets

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 10]
```diff
@@ -135,53 +206,57 @@ List<EnumValueDescriptor> parseEnumValues(List<?> enums){
...
         }
     }
 
-    <D extends PropertyDescriptor> D parseDescription(Schema<?> schema, D propertyDescriptor){
-
-        if (schema.getDescription() != null)
-            propertyDescriptor.setDescription(schema.getDescription());
-
-        return propertyDescriptor;
+    private TypeDescriptor parseXof(Schema x){
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:246]: maybe choose a more speaking name than "x"

File: src/test/java/org/jqassistant/plugin/openapi/parser/JsonSchemaTest.java
Comment (CodeOfRob) [inline src/test/java/org/jqassistant/plugin/openapi/parser/JsonSchemaTest.java:1]: test coverage looks good, would wish additional coverage of `schema.getOneOf`and schema.getAnyOf`.

am skipping detailed review of the tests and trusting sonars analysis

**Review by CodeOfRob — CHANGES_REQUESTED** [review]
looks really good, i think i found some small mistakes. pls take a look at my comments

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java (hunk 1 above)
Comment (jp06upel) [reply src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:59]: yes this function always parses/overwrites! to get a reference, you should use the ResolverClass.

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java (hunk 5 above)
Comment (jp06upel) [reply src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:68]: woops copy pasta hit again xD

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java (hunk 2 above)
Comment (jp06upel) [reply src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:75]: i dont quite get what you mean here.

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java (hunk 8 above)
Comment (jp06upel) [reply src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:142]: yep but i would like to create a new ticket for this.

Comment (UzeyirMammadov) [reply src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:142]: > yep but i would like to create a new ticket for this.

Agree, it's out of scope if this ticket

File: src/main/java/org/jqassistant/plugin/openapi/impl/parsers/ComponentsParser.java [hunk 11]
```diff
@@ -18,6 +18,8 @@ public class ComponentsParser {
     public static ComponentsDescriptor parse(Components components, Store store){
         ComponentsDescriptor componentsDescriptor = store.create(ComponentsDescriptor.class);
 
+        // INSTANZ
+
```
[the remark this answers is not in the document]
Comment (jp06upel) [reply src/main/java/org/jqassistant/plugin/openapi/impl/parsers/ComponentsParser.java:22]: removed (i dont remember adding this comment)

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java (hunk 2 above)
Comment (CodeOfRob) [reply src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:81]: with every condition you check, a new possible combination of conditions is created, which needs to be tested.

> if( con1 ) --> two possible branches to test
> if( con1 && con2) --> four possible branches to test

by decreasing the condition count you decrease the combinations to test and increase the coverage KPI (with the same tests)

also, a for loop should be able to process empty lists without throwing an error. So the .isEmpty() check is not quite necessary

Comment (jp06upel) [reply src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:81]: got ya!

File: src/test/resources/example-jsonschema.yaml [hunk 12]
```diff
@@ -74,3 +79,43 @@ components:
...
+              type: string
+    SchemaWithInherentOneOf:
+      oneOf:
+        - $ref: '#/components/schemas/SchemaWithObject'
+        - type: object
+          properties:
+            name123:
+              type: string
+    SchemaWithInherentAnyOf:
+      oneOf:
```
Comment (CodeOfRob) [inline src/test/resources/example-jsonschema.yaml:97]: image

File: src/test/java/org/jqassistant/plugin/openapi/parser/JsonSchemaTest.java [hunk 13]
```diff
@@ -219,11 +264,86 @@ void schemaWithReferenceTypeTest(){
...
+        for (TypeDescriptor type : allOffs){
+            assertThat(type).isInstanceOfAny(ReferenceTypeDescriptor.class, ObjectTypeDescriptor.class);
+        }
+    }
+
+    @Test
+    void SchemaWithInherentAnyOf(){
+        SchemaDescriptor schemaDescriptor = getSchemaWithName("SchemaWithInherentAnyOf");
+
+        List<TypeDescriptor> allOffs = schemaDescriptor.getOneOfSchemas();
```
Comment (CodeOfRob) [inline src/test/java/org/jqassistant/plugin/openapi/parser/JsonSchemaTest.java:314]: image

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 14]
```diff
@@ -135,53 +212,56 @@ List<EnumValueDescriptor> parseEnumValues(List<?> enums){
...
-    IntegerPropertyDescriptor parseInteger(final Schema<?> schema){
+    private IntegerTypeDescriptor parseInteger(final Schema<?> schema){
 
-        return parseDescriptionAndFormat(schema, store.create(IntegerPropertyDescriptor.class));
+        return parseFormat(schema, store.create(
+                IntegerTypeDescriptor.class));
     }
 
-    NullPropertyDescriptor parseNull(final Schema<?> schema){
+    private NullTypeDescriptor parseNull(final Schema<?> schema){
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:221]: sonar wants the unused parameter to be removed

File: src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java [hunk 15]
```diff
@@ -15,115 +16,191 @@
...
-        arrayPropertyDescriptor.setItem(parseProperty(schema.getItems(), "items"));
+        ArrayTypeDescriptor arrayTypeDescriptor = store.create(ArrayTypeDescriptor.class);
+        arrayTypeDescriptor.setItem(parseType(schema.getItems()));
 
-        return arrayPropertyDescriptor;
+        return arrayTypeDescriptor;
     }
 
-    BoolPropertyDescriptor parseBoolean(final Schema<?> schema){
+    private BoolTypeDescriptor parseBoolean(final Schema<?> schema){
```
Comment (CodeOfRob) [inline src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:189]: sonar wants the unused parameter to be removed

Comment (UzeyirMammadov) [reply src/main/java/org/jqassistant/plugin/openapi/impl/jsonschema/JsonSchemaParser.java:189]: test

File: src/test/java/org/jqassistant/plugin/openapi/parser/JsonSchemaTest.java (hunk 13 above)
Comment (CodeOfRob) [reply src/test/java/org/jqassistant/plugin/openapi/parser/JsonSchemaTest.java:314]: is this really intended or did you want to `schemaDescriptor.getAnyOfSchemas()` here?

**Review by CodeOfRob — APPROVED** [review]
well done 🥳

**Review by CodeOfRob — APPROVED** [review]
